### ISSUE TITLE
Ship plugin manager with knossos

### ISSUE THREAD

**Issue by jmrk84** [knossos-project/knossos#136]
.... it should be added to the resource files and there should be a menu entry (in the python pull down) that allows launching it directly. Not yet sure how the plugin manager can then be updated, but it is way too annoying to download pm.py from somewhere and then launch it manually every time.

**Comment by jmrk84** [comment]
I think it makes more sense to not include the plugin manager as a resource file, but to inlcude a mini-python plugin as a resource file that automatically downloads the latest version of the plugin manager when clicking on the menu item from the repository. This way, the plugin manager itself can never become outdated. After the first download, the plugin manager file could also be stored locally and updated.

**Comment by Optiligence** [comment]
You don’t need a plugin to download something…
If the plugin manager can update itself though, it could be embedded and extracted to the plugin directory if it’s not already present and then update itself.

**Comment by jmrk84** [comment]
We could have a plugin to download the plugin downloader... We would have many plugins soon! But I agree, your proposal sounds much better.

**Comment by orenshatz** [comment]
What exactly *is* the proposal? The ability to self-update the plugin manager would have been required regardless of another mini-downloader, and I took this ability for granted. The python file is not read-only. The only problem is that the instance that was created is still in memory, and this would never be solved, if the user updates the plugin manager he should just restart knossos. We can write it in the plugin's description or in another metadata. But this is general knowledge to be applied to all plugins. Unless we want a mechanism to identify a plugin's name with the class+instance and deleting them upon update, of course this would require special handling in case the plugin currently runs, and I suggest to drop it.

**Comment by jmrk84** [comment]
The proposol is to embed the plugin manager and extract it to the plugin folder from C++ the first time it is selected in the python pull down menu. Really, in the end it makes no big difference how exactly this happens (C++ or python code, ...), but the current situation of downloading the file manually and running the (a bit cryptically named) pm.py is not sufficient.

**Comment by Optiligence** [comment]
I would also like the have the non-restart plugin update. 
> the plugin currently runs
If you have unsaved data within the plugin, restarting Knossos would also loose it.

**Comment by orenshatz** [comment]
Joergen, but to embed the plugin manager as a resource is what this issue is all about to begin (I have no idea how resources are opened, obviously if it's not part of the installer then some C++ code would have to do that), am I missing something?

Norbert, I assume you refer to updating not only the plugin manager with restart, but any plugin. How do you suggest synchronizing it with the running plugin, without requiring something from plugin authors?

**Comment by Optiligence** [comment]
If the plugins don’t reference each other, you should be able to just delete the plugin in question and replace it with the updated one, no?
Resources are easy…

**Comment by orenshatz** [comment]
First you need to reference the plugin and its instances. But this can be solved by the global plugin_container serving as a named hash (then we should require plugin authors to conform to this hashing instantiation scheme).
But more importantly, if a plugin is currently running and the plugin manager wants to update it, then the running plugin has to be notified, unless of course the user is aware of this and stops the plugin, in a way that also frees any resources held by it (let's assume plugin authors implement the __del__ on their instances, and that plugins are never imported).
Which leaves us with the special case of plugin manager itself, which would need a transiently-acting help class in order to self-reboot.

Joergen, confirm if you support the above.

**Comment by Optiligence** [comment]
The thing is, if the plugins don’t reference each other, there is no difference deleting the running plugin instance and closing Knossos, so there’s no need for a delete handler…
If we want a general exit Handler, this would run on deletion as well as K restart.

**Comment by jmrk84** [comment]
I have never worked with resources myself, but it looks like Norbert can help you with the embedding. Runtime updating is nice, but we have much more important stuff to do.

**Comment by orenshatz** [comment]
Alright, to future-support awareness of the plugin manager to other running plugins, I will already introduce tha hashing scheme for plugin classes and instance in plugin manager itself, as reference for other plugin writers, just so we don't have to ask them to change their code later to support this.
And otherwise, I will just pack pm.py in the resources, unpack it at run-time to the default plugin dir (redundant default dir creation code with the pm but anyway) and hardcode a button to run it from the python menu.
Okay?

**Comment by Optiligence** [comment]
When Knossos starts you want to find out if the plugin manager exists already. 
(Do we want to support several plugin directories or only the default one?)
(Shouldn’t the default location be in AppData like the other Knossos files, reachable through a menu entry within Knossos and have a better name, like ›Knossos Plugins‹)
If it doesn’t exist or if the version you’d put there is newer, replace it. (either through Resources or Online)
And a menu item, of course.

**Comment by orenshatz** [comment]
I'm not going to introduce C++ code that parses the plugin manager to find out if it's newer,
I don't care if it is, the entire resource unpack routine would only run once upon the first knossos
launch (and then mark a runonce flag in QSettings).
The menu item I already wrote above.
(plugin directories on top of the default ones can be supported later by pm, no need for it now.
I keep the filesystem 8.3 compatible as convention, hence the name, which no user should ever
care about as long as pm serves him right.
I chose the user env home because plugins are per-user, I don't know how AppData behaves
on OS other than Windows.)

**Comment by Optiligence** [comment]
You gotta find out somehow if there is an update available. 
AppData, .config, .local/share is always per User and i think it’d be nice if all’d stay in the same location. 

I would run it only if the pm is not already present and not if some flag is not set. If someone deletes the pm he would’ve to unset the flag too, which is unfortunate. 
> plugin directories on top of the default ones can be supported later by pm, no need for it now

This would mean, that the pm always stays in the default location and the separate plugin locations would be transparent to C++ Knossos. On could’ve also provided the plugin location (including pm) as a variable inside Knossos. 

> I keep the filesystem 8.3 compatible as convention

Really? I’d rather break stuff for people in the beginning than later. It just enforces good/future-proof code and ›Knossos Plugins‹ also reads nicer.

**Comment by orenshatz** [comment]
Why would someone delete pm if he wants it later?
Okay anyway, AppData (I hope you know it's per-user for Mac), no run-once, always create if missing on startup, but forget the overwrite if older or the update-check, this would be handled by the pm itself when the responsible user runs it, unless we want a non-GUI pm that can update itself silently.

Yes, non-default python stuff are generally not accessible from C++, that's obvious.
Whatever is non-default that we want C++ to support would have to be coded in C++,
but for now we don't know of any requirement for non-default plugin dirs. How would such
variable be supported anyway, in QSettings?

Why break anything if unncecessary, nothing proof about a longer nice name than shorter ugly.

**Comment by Optiligence** [comment]
> How would such variable be supported anyway, in QSettings?

Yes, a plugin path variable with the default path as default value. (but yeah, i neither know of a use case^^)

> Why break anything if unncecessary, nothing proof about a longer nice name than shorter ugly.

Why would you use Unicode (#82), it’s the ›why don’t we still live in a cave‹. 
The plugin creators are educated, at least i envision it that way.

**Comment by orenshatz** [comment]
Unicode is used by the user, the default dir is used by us, the user suffers no consequence.

**Comment by Optiligence** [comment]
apart from looking nicer, i would use the space in ›Knossos Plugins‹ to let people see if their programs don’t support it, so they fix it
i also don’t get your last message at all

**Comment by orenshatz** [comment]
What do you mean "their programs", I just want people to be able to place the plugins on a 8.3 filesystem if necessary,
Bug #82 is about a user-selected username, that's a constraint placed by us that limits ther user,
therefore a bug. This is not the case here, here the user doesn't select anything, we select.

**Comment by Optiligence** [comment]
> What do you mean "their programs", I just want people to be able to place the plugins on an 8.3 filesystem if necessary,

And i would like to utilize progress in cs (at least for the default location). 

> Bug #82 is about a user-selected username, that's a constraint placed by us that limits ther user,
therefore a bug. This is not the case here, here the user doesn't select anything, we select.

It’s a bug discovered because a user used an ›unusual‹ username. That’s the the thing i want to accomplish here, let plugin developers discover bugs…

**Comment by orenshatz** [comment]
"progess in cs"?
Which bug are *they* going to discover by *us* using a space?
If someone wants plugins stored on an 8.3 filesystem, it is *no* bug.

**Comment by Optiligence** [comment]
But if they develop plugins which can’t handle spaces it is. 
Point being: i think ›knos_plg‹ is a really ugly choice. 
And just a note regarding user folder and AppData: if you run Knossos from QtCreator it uses the msys home directory…

**Comment by orenshatz** [comment]
What do mean "plugins which can't handle spaces", I'm talking about the filenames...
Ugly or not, it works, and the user should never care about the name.

**Comment by fabiansvara** [comment]
I think it will be very hard to find a computer that will run Knossos and have an 8.3 limit on its file system at the same time. Just use something, anything, human readable.

**Comment by fabiansvara** [comment]
As to the plugin update issue: I agree with Jörgen that having actualy plugins should be the main concern currently. But why is this an issue at all? Re-evaluating class and function definitions should just work.

**Comment by fabiansvara** [comment]
One thing we need to do is define a "plugin protocol". This is up for discussion. My working suggestions: Every plugin should be a python package, so as to minimize conflicts in the global namespace and organize plugins in a standardized fashion. This would also require a defined entry-point for execution, e.g. a run() function (not method, to allow for programming styles without classes) in a main.py file. Additionally, there should be some package-global flags, such as VERSION or RUN_ON_LOAD. The latter would control whether the run() method is called on plugin load. In order to support plugins calling each other's code, there must be a mechanism for plugins to ensure that a different plugin is loaded. This should not happen through simple importing to make sure that the plugin manager knows about loaded plugins. Instead, we could support a plugin_manager.load_plugin() method that imports a package and registers it with the plugin manager.

**Comment by fabiansvara** [comment]
I have sort of changed my mind on plugin reloading. I think we should forget about it and not support it because we shouldn't make this assumption

> The thing is, if the plugins don’t reference each other,

It would be a good thing for plugins to reference each other. The way out would be for the plugin manager to call reload() on all plugin packages when one plugin changes, which might have all sorts of other side effects. We might consider simply adding a "Reload Plugins" button in the future, but I do not see this as a priority.

**Comment by orenshatz** [comment]
There are too many suggestion floating around here, we cannot ship anything, including plugin manager itself, before they are resolved.
To what Fabi said, I don't get exactly what a "package" is, in terms of actual evaluation in python and representation in files, and "import". One thing is sure, you cannot import twice. And as I said above, you can't also just "parse again" a class to replace the old one, if the old one is still running or has leftovers (instances or produced data structures) living as references by other classes. We should be very clear about the responsibility of the user to clean for himself and update for himself etc, the plugin manager is as dumb as can be. The run function of the package (why not a class, there are already unbound methods, and the class encapsulation is good enough for the namespace isolation) can be called or not simply by *running* it or not at the end of the python file.
I currently require *no* python evaluation to read metadata about a plugin, by using simple magic-header python comments. This avoid crashing on bugs, or actually loading the plugin, when the user only wanted to get an update.
About integration into knossos menus, this should first be supported by interfacing MainWindow and it's members to python.

**Comment by fabiansvara** [comment]
Packages: https://docs.python.org/2/tutorial/modules.html#packages

When python sees import package_name, it executes package_name/__init__.py. Conventionally, __init__.py will then import from modules and sub-packages within the package, making their names available in the global context as package_name.package_local_name.

Indeed, you can not import twice. This is what reload (https://docs.python.org/2/library/functions.html#reload) is for. Reload will not get rid of references that still exist in other plugins, for example. Which is why I was suggesting a "Reload all Plugins" button. But it would be important to clarify how this affects python code that is already a slot for some signal. The complexity of this is what leads me to believe that just restarting Knossos is less hassle for everyone.

> The run function of the package (why not a class, there are already unbound methods, and the
> class encapsulation is good enough for the namespace isolation)

Using packages will avoid module name clashes. Simply grouping functions in separate modules without using classes is a common python idiom. Also, using packages and the standard behavior of __init__.py will offer the least surprise for those familiar with python, and there will be no need to define anything Knossos specific.


> About integration into knossos menus, this should first be supported by interfacing MainWindow
> and its members to python.

Ok, this is important then. The reason I was talking about auto-loading (or rather: auto-importing of packages) and about plugins inserting themselves into the GUI on import is that in the long-term I expect the standard Knossos distribution to ship with many plugins, and it should be irrelevant and transparent to the casual user whether a given feature is implemented in C++ or Python. This also means that plugins should be able to easily extend all or most parts of the Knossos GUI and keyboard shortcuts, so manually starting individual plugins through the plugin manager should not be required for most plugins.

**Comment by jmrk84** [comment]
I fully support the package idea. This way we can avoid ugly name space collisions and plugins can easily make use of functionality in other plugins. Addtionally, we should keep the "run .py file" button that does simply that, without any checks etc, to allow devs to shoot in their foot. But most important: We need the example plugins (apart from the pm). Already implemented as packages ideally. I will open an issue for each of the plugins I think that would be nice (and still easy to do).

**Comment by fabiansvara** [comment]
One more thing we need it for the plugins to specify a minimum knossos version they require.

**Comment by orenshatz** [comment]
What do you mean by "need"? They can simply decide not to run.
What we do need is a python interface to the knossos version, which I'll add now.

**Comment by Optiligence** [comment]
tja, too late now

**Comment by fabiansvara** [comment]
Spending the last two hours or so fiddling with os.chdir() and sys.path in the only-somewhat-userfriendly Knossos Python shell has reminded me that we *really* need a standardized way of shipping plugins with an auto-import (Python import keyword) mechanism.

**Comment by orenshatz** [comment]
Could have spent less time using the python path and python working dir features of knossos python configuration widget instead.

**Comment by fabiansvara** [comment]
Having to think about them at all is wrong.

**Comment by orenshatz** [comment]
I don't see the point, you have to place your plugin *somewhere*, or else knossos wouldn't know where it is. if they're in the python path, you can just import them. if they're in a custom auto-import path (whether post-distribution or at build-time), you have to be aware of that specific path. once you're aware of it, you can simply define this path in the python-path field of the python configuration widget as a temporary work-around. in the future, auto-import would be defined in the plugin as a metadata field, and the plugin would be introduced either at build time by placement in the resources/plugin dir, or post-build by manual placement in the local plugins dir (that the user would have to be aware of) or in the server-side repository. so the user would have to think about *something* no matter how we work it out.

**Comment by Optiligence** [comment]
I think it would be good to know what fabi tried to do exactly and what didn’t work but should have.
